portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -58,21 +76,46 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get name and text
-    String name = request.getParameter("name-input");
-    String text = request.getParameter("text-input");
-    if (text != null && name != null) {
-      long timestamp = System.currentTimeMillis();
-
-      Entity messageEntity = new Entity(Constants.MESSAGE_ENTITY_TYPE);
-      messageEntity.setProperty("name", name);
-      messageEntity.setProperty("text", text);
-      messageEntity.setProperty("timestamp", timestamp);
-
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-      datastore.put(messageEntity);
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+        String userEmail = userService.getCurrentUser().getEmail();
+        
+        // Get name and text
+        String name = getUserNickname(userService.getCurrentUser().getUserId());
+        String text = request.getParameter("text-input");
+        if (text != null && name != null) {
+        long timestamp = System.currentTimeMillis();
+
+        Entity messageEntity = new Entity(Constants.MESSAGE_ENTITY_TYPE);
+        messageEntity.setProperty("name", name);
+        messageEntity.setProperty("text", text);
+        messageEntity.setProperty("email", userEmail);
+        messageEntity.setProperty("timestamp", timestamp);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(messageEntity);
+      }
     }
   }
+
+  /**
+   * Returns the nickname of the user with id, or empty String if the user has not set a nickname.
+   */
+  private String getUserNickname(String id) {
```

Comment:
So this is a fun question that we have to deal with all the time in drive.

As is: if a user changes their nickname, they could leave comments that appear to come from multiple people. How could you fix that? (This is a hard question, and more of an exercise).

Reply:
When they change their nickname I can load in data, check if it came from the same email and then change the name. But I can see how it would get hard/resource-intensive if there are multiple db where the name could have been stored.

Maybe have one db that stores the email and name. That way if we load in data we can use the email to reference the name.

I just looked at my code and realized I can store the comment with a user Id and retrieve the nickname on load. I will try that when I look into implementing Objectify.